include/sys/dmu.h

```diff
@@ -797,9 +800,11 @@ int dmu_write_uio_dnode(dnode_t *dn, struct uio *uio, uint64_t size,
 #endif
 struct arc_buf *dmu_request_arcbuf(dmu_buf_t *handle, int size);
 void dmu_return_arcbuf(struct arc_buf *buf);
-void dmu_assign_arcbuf(dmu_buf_t *handle, uint64_t offset, struct arc_buf *buf,
-    dmu_tx_t *tx);
-void dmu_assign_arcbuf_impl(dmu_buf_t *handle, struct arc_buf *buf,
+void dmu_assign_arcbuf(objset_t *os, uint64_t object, uint64_t offset,
+    struct arc_buf *buf, dmu_tx_t *tx);
```

Comment:
In order to avoid needlessly breaking the Lustre build, which depends on this symbol, let's do the following. 

* Rename `dmu_assign_arcbuf()` to `dmu_assign_arcbuf_by_dbuf()` as you've already done.  This is a better name and consistent with the other `_by_dnode/_by_dbuf` changes.
* Add a #define in the header to map the `dmu_assign_arcbuf` function to `dmu_assign_arcbuf_by_dbuf`.
* Update the one call site where you needed the new `dmu_assign_arcbuf()` to take the dnode hold itself so you don't need this wrapper.
* Concurrently let's give @adilger and @bzzz77 a heads up so we they can migrate to the new `_by_dbuf` interface.
* At some point in the future we can then retire the compatibility code and add the alternate version of `dmu_assign_arcbuf` proposed here.